server/src/main/java/com/cloud/template/TemplateManagerImpl.java

```diff
@@ -616,6 +616,18 @@ public void prepareIsoForVmProfile(VirtualMachineProfile profile, DeployDestinat
 
     private void prepareTemplateInOneStoragePool(final VMTemplateVO template, final StoragePoolVO pool) {
         s_logger.info("Schedule to preload template " + template.getId() + " into primary storage " + pool.getId());
+        if (pool.getPoolType() == Storage.StoragePoolType.DatastoreCluster) {
+            List<StoragePoolVO> childDataStores = _poolDao.listChildStoragePoolsInDatastoreCluster(pool.getId());
+            s_logger.debug("Schedule to preload template " + template.getId() + " into child datastores of DataStore cluster: " + pool.getId());
+            for (StoragePoolVO childDataStore :  childDataStores) {
+                prepareTemplateInOneStoragePoolInternal(template, childDataStore);
+            }
+        } else {
+            prepareTemplateInOneStoragePoolInternal(template, pool);
+        }
+    }
+
+    private void prepareTemplateInOneStoragePoolInternal(final VMTemplateVO template, final StoragePoolVO pool) {
```

Comment:
naming is kind of off; `private prepareTemplateInOneStoragePool(..)` is already internal as the fact that it is `private` suggests. I would think this name would point to what it actually does, like `startAThreadToPreloadTemplate(..)` or such a thing.